src/logging.rs

```diff
@@ -83,52 +83,79 @@ fn add_field_and_payload(data: &mut String, field: &str, payload: &str) {
     }
 }
 
-/// Print a message to the systemd journal with the given priority.
+/// Print a message to the journal with the given priority.
+///
+/// This uses the default systemd-journald socket.
 pub fn journal_print(priority: Priority, msg: &str) -> Result<()> {
     let sock = UnixDatagram::unbound()?;
 
     let mut data = String::new();
     add_field_and_payload(&mut data, "PRIORITY", &(u8::from(priority)).to_string());
     add_field_and_payload(&mut data, "MESSAGE", msg);
 
-    let res = sock.send_to(data.as_bytes(), SD_JOURNAL_SOCK_PATH);
-    match res {
-        Ok(_) => return Ok(()),
-        // If error code is 90, the message was too long for a UNIX socket.
-        Err(ref e) if e.raw_os_error() == Some(90) => {
-            let tmpfd = memfd_create(&CString::new("journald")?, MemFdCreateFlag::MFD_ALLOW_SEALING)?;
-            // Safe because memfd_create gave us this FD.
-            let mut file = unsafe { File::from_raw_fd(tmpfd) };
-            file.write_all(data.as_bytes())?;
-
-            let memfd = file.into_raw_fd();
-            fcntl(memfd, FcntlArg::F_ADD_SEALS(SealFlag::all()))?;
-            let fds = &[memfd];
-            let ancillary = [ControlMessage::ScmRights(fds)];
-            let path = SockAddr::new_unix(SD_JOURNAL_SOCK_PATH)?;
-            sendmsg(sock.as_raw_fd(), &[], &ancillary, MsgFlags::empty(), Some(&path))?;
-        },
-        _ => return Err("unknown err".into()),
+    // Message sending logic:
+    //  * fast path: data within datagram body.
+    //  * slow path: data in a sealed memfd, which is sent as an FD in ancillary data.
+    //
+    // Maximum data size is system dependent, thus this always tries the fast path and
+    // falls back to the slow path if the former fails with `EMSGSIZE`.
+    let fast_res = sock.send_to(data.as_bytes(), SD_JOURNAL_SOCK_PATH);
+    let res = match fast_res {
+        // `EMSGSIZE` (errno code 90) means the message was too long for a UNIX socket,
```

Comment:
Thanks for documenting this, definitely not clear before what was going on here!